File: softgoals/nfr3.py
#!/usr/bin/env python3 -B
# nfr3.py : nfr3.pl said in python. The model is data and the
# DSL is operator abuse: nodes spring into existence when
# mentioned (__getattr__), bodies conjoin with `&`, or-clauses
# accumulate with `|=` (each clause IS a disjunct, so the
# operators mean what logic says). As in nfr3.pl, hard vs soft
# is read off body shapes: any all-hard body makes a rule
# node (choice-or, commit), else an edge node (label all
# bodies, combine). ISAMP replaces backtracking: a
# contradiction raises, the world restarts fresh (nfr3's greedy
# mode is the only mode).
# A Node is a logic Var with bodies attached (head |= body
# asserts one clause; the head is the Node itself). A Var
# with no bodies is an abducible: assumable, never proved
# (ALP framing: theory + abducibles -> explanations).
# Var.binding is None (unbound) or a number. Head-first, like
# the prolog: a rule binds 2 then walks its body; an edge binds
# a coinductive guess (uniform over the 5 labels) then walks;
# either way a loop-back just reads a bound var -- a consistent
# subset -- and the edge's own combine verifies guess ==
# computed, else the world restarts. unify is prolog unify
# grown one abductive habit: asked to bind with no demand it
# guesses (labeling): abducibles stagger +-2.
# Labels 2 1 0 -1 -2. One looseness vs the prolog: python
# cannot close the world, so a typo is a new leaf, not an
# existence error.
import random

class Fail(Exception): pass

def damp(v): return max(-1, min(1, v))

def shuffled(xs): return random.sample(list(xs), len(xs))

class Lit:                      # `&` conjoins literals into a
  def __and__(i, o): return [i, o]        # clause (a list);
  def __rand__(i, o): return o + [i]      # rand extends one

class Term(Lit):                # no/makes/.../And/Or wrapper
  def __init__(i, op, x): i.op, i.x = op, x

def no(x):     return Term('no', x)
def makes(x):  return Term('makes', x)
def breaks(x): return Term('breaks', x)
def helps(x):  return Term('helps', x)
def hurts(x):  return Term('hurts', x)
def And(*xs):  return Term('amin', xs)
def Or(*xs):   return Term('amax', xs)

def hardlit(l): return isinstance(l, Node) or l.op == 'no'

class Var(Lit):
  "k : name; binding : None | label"
  def __init__(i, k): i.k, i.binding = k, None
  def unify(i, want=None, vals=(2, -2)):
    v = i.binding
    if v is None:                     # unbound: bind the
      i.binding = v = want if want is not None \
                           else random.choice(vals)  # demand,
      return v                        # else guess (labeling)
    if want is None or v == want: return v  # bound: agree
    raise Fail                              # or die

class Node(Var):
  "a Var plus its bodies, split hard/soft as they arrive"
  def __init__(i, k): Var.__init__(i, k); i.hards, i.softs=[],[]

  def abducible(i): return not (i.hards or i.softs)

  def __ior__(i, b):            # head |= body (an or); any
    b = b if isinstance(b, list) else [b]   # all-hard body
    (i.hards if all(hardlit(l) for l in b)  # makes a rule
             else i.softs).append(b)        # node, else edge
    return i

  def combine(i, ws):
    hi, lo = max([0] + ws), min([0] + ws)
    if hi == 2 and lo == -2: raise Fail  # sat meets denied
    return i.unify(hi + lo)     # computed must match my guess

  def __call__(i, want=None):
    if i.binding is not None: return i.unify(want)  # memo
    if i.abducible(): return i.unify(want)    # stagger +-2
    if i.hards:                 # rule: pick one body, walk
      if want == -2: raise Fail # it; falsity is assumed,
      i.binding = 2             # never derived (soft bodies
      for l in shuffled(random.choice(i.hards)):  # dead)
        l(2) if isinstance(l, Node) else l.x.unify(-2)
      return 2
    if want is not None:        # edge: a demand is assumed,
      i.binding = want          # not derived (nfr2 guess
      return want               # parity)
    i.binding = random.choice((2, 1, 0, -1, -2))  # guess,
    es = [e for c in i.softs for e in c]            # head-first
    return i.combine([contrib(e) for e in shuffled(es)])

class Theory:                   # name nodes reserved. `m.x |= b`
  def __init__(i): i.__dict__['nodes'] = {}   # reads m.x (making
  def __getattr__(i, k): return i.nodes.setdefault(k, Node(k))
  def __setattr__(i, k, v): i.nodes[k] = v    # it) then writes
                                              # it back: same Node

# ---- soft side: contributions are plain numbers -----------
def contrib(e):
  if isinstance(e, Node): return e()        # bare = makes
  if e.op == 'no':   return e.x.unify(-2)
  if e.op == 'amin': return min(contrib(x) for x in e.x)
  if e.op == 'amax': return max(contrib(x) for x in e.x)
  v = e.x()
  return {'makes': v,        'breaks': -v,
          'helps': damp(v),  'hurts': -damp(v)}[e.op]

# ---- top: a world = one try, 100 restarts max -------------
def world(m, f):
  for _ in range(100):
    for n in m.nodes.values(): n.binding = None
    try: return f()
    except Fail: pass

def picks(m):
  out = []
  for n in m.nodes.values():
    if n.binding is not None and n.abducible():
      out.append(n.k if n.binding == 2 else
                 ('no', n.k) if n.binding == -2 else
                 ('lab', n.k, n.binding))
  return tuple(sorted(out, key=str))

def abduce(m, g):               # one world proving g, shown
  def run(): g(2); return picks(m)         # as its leaf picks
  return world(m, run)

def soften(m, g):               # one world labelling g
  def run(): return g(), picks(m)
  return world(m, run)

def worlds(n, f):               # sample n, keep distinct
  ws = {w for w in (f() for _ in range(n)) if w}
  return sorted(ws, key=str)

# ---- keys: score worlds, credit settings, find plateau ----
def settle(m, hs, ss, priors=()):  # one world under priors:
  vs = list(m.nodes.values())      # score, and its +-2
  xs = [n for n in vs if n not in ss]        # settings
  for _ in range(100):
    for n in vs: n.binding = None
    for n, v in priors: n.binding = v
    try:
      for h in shuffled(hs): h(2)
    except Fail: continue
    good = 0
    for s in shuffled(ss):
      snap = [(n, n.binding) for n in vs]
      try:
        if s() >= 2: good += 1
      except Fail:
        for n, b in snap: n.binding = b
    used = sum(n.binding is not None for n in xs)
    return (good/len(ss) - used/len(xs),   # max soft, min use
            [(n, n.binding) for n in vs
             if n.abducible() and n.binding in (2, -2)])

def keys(m, hs, ss, runs=100, reps=20, same=None):
  credit = {}                # setting -> scores of its worlds
  for _ in range(runs):
    if r := settle(m, hs, ss):
      for kv in r[1]: credit.setdefault(kv, []).append(r[0])
  mid = lambda a: sorted(a)[len(a)//2]
  rank = sorted(credit, key=lambda kv: -mid(credit[kv]))
  score = lambda pr: sorted(r[0] for _ in range(reps)
                            if (r := settle(m, hs, ss, pr)))
  prev, k = score(()), 0     # assert top-k until no gain
  while k < len(rank):
    cur = score(rank[:k+1])
    if mid(cur) <= mid(prev) or same and same(prev, cur): break
    prev, k = cur, k + 1
  return rank[:k], prev

# ---- eg: the nfr3-eg.pl graphs ----------------------------
m = Theory()
m.happy |= m.rich
m.happy |= m.loved & no(m.lonely)
m.rich  |= m.works & m.lucky
m.loved |= m.friends
m.friends |= m.happy            # loop
m.g |= m.p & m.q
m.p |= m.x
m.q |= no(m.x)                  # x and no(x): no worlds
m.performance |= helps(m.indexing) & hurts(m.logging)
m.security    |= makes(m.encryption) & hurts(m.indexing)
m.usability   |= breaks(m.encryption) & helps(m.gui)
m.good  |= And(m.performance, m.security, m.usability)
m.trust |= helps(m.good) & makes(m.trust)  # self loop
m.hard_goals |= m.happy
m.soft_goals |= Or(m.performance, m.security, m.usability)

if __name__ == '__main__':
  w = lambda n, g: worlds(n, lambda: abduce(m, g))
  print("happy", w(300, m.happy))
  print("g    ", w(100, m.g))
  print("hard ", w(300, m.hard_goals))
  ws = worlds(600, lambda: soften(m, m.good))
  print("good ", len(ws), "worlds, best",
        max(ws, key=lambda x: x[0]))
  print("trust", sorted({s[0] for s in
        (soften(m, m.trust) for _ in range(300)) if s}))

File: softgoals/nfr3.pl
% nfr3.pl : nfr2.pl without the meta-interpreter. `<--` clauses
% stay in the database as data; a preprocess/0 pass compiles each
% head into real clauses (memo clause first, so loops share a value
% instead of recursing; then believe-head-and-walk-body), and every
% leaf gets a maybe/4 stub. and/or/no and the soft
% contributions (makes/breaks/helps/hurts) are ordinary predicates
% threading a belief list DCG-style, values passed up as an extra
% argument: hard ops are arity 3 (no value: a positive literal
% demands 2), soft ops arity 4. The or-policy is still picked by
% the defining database: any all-hard body makes the head a rule
% node (choice-or, edges dead, as nfr2's agenda cut); else an edge
% node (label ALL bodies, combine). nfr2 parity is exact, including
% its quirk that a hard demand on an edge node is assumed, not
% derived (see the guess clause in nfr2's eval//2).
:- op(1100, xfx, <--).
:- op(900,  fy,  no).
:- dynamic (<--)/2, greedy/0.        % greedy: commit or-choices
:- discontiguous (<--)/2.

permute(Xs,Ys) :- random_permutation(Xs,Ys).

% ---- belief set: a Node-Value list threaded as DCG state -----------------
peek(S,S,S).
push(X,S,[X|S]).
=(X,X,S,S).

believed(K,V,S,S) :-  memberchk(K-V0,S), V = V0.

% maybe(K,V): recall K's value, else assume one (undone on backtrack).
% Callers cut on believed with a FRESH var, unify after: commits to
% the recalled value, then a mismatched demand fails outright.
maybe(K,V,S,S) --> believed(K,W,S,S), !,  V = W.
maybe(K,V) --> { var(V) -> permute([2,-2],Ps), member(V,Ps) ; true }, push(K-V).

% ---- hard side: or//1, and//1, no//1 -------------------------------------
% falsity is assumed, never derived: `no` never runs its node's clauses.
no(K) --> maybe(K,-2).

or(Bs) --> { greedy, !, random_member(B,Bs) }, and(B).       % ISAMP: no retry
or(Bs) --> { permute(Bs,Rs), member(B,Rs) }, and(B).

and(Ls) --> { permute(Ls,Rs) }, ands(Rs).
ands([])     --> [].
ands([L|Ls]) --> lit(L), ands(Ls).

lit(no(K)) --> !, no(K).
lit(K)     --> call(K,2).

% ---- soft side: contributions are nonterminals too -----------------------
makes(X, V)            --> call(X,V).                        % full, same
breaks(X,neg(W))       --> call(X,W).                        % full, flip
helps(X, damp(W))      --> call(X,W).
hurts(X, neg(damp(W))) --> call(X,W).
and(Xs,  amin(Vs))     --> calls(Xs,Vs).
or(Xs,   amax(Vs))     --> calls(Xs,Vs).
no(X,    -2)           --> no(X).              % assume false, contribute -2

calls([],[])         --> [].         % one fold serves edge lists and the
calls([X|Xs],[V|Vs]) --> call(X,V), calls(Xs,Vs).            % and/or args

% contributions arrive symbolic (a loop label may still be a var);
% ground the pendings, then evaluate the expressions.
evalx(X,X)       :- number(X), !.
evalx(neg(E),V)  :- evalx(E,W), V is -W.
evalx(damp(E),V) :- evalx(E,W), V is sign(W)*min(1,abs(W)).
evalx(amin(L),V) :- maplist(evalx,L,Ws), min_list(Ws,V).
evalx(amax(L),V) :- maplist(evalx,L,Ws), max_list(Ws,V).

combine(Vs, V) :- term_variables(Vs, Us), grounds(Us),  % pending loop labels
                  maplist(evalx, Vs, Ws),
                  max_list([0|Ws], P),                 % strongest support,
                  min_list([0|Ws], N),                 % strongest objection
                  \+ (P =:= 2, N =:= -2),              % sat meets denied
                  V is P + N.

grounds([]) :- !.
grounds(Us) :- length(Us,N), N > 3, !,       % big cyclic cluster: punt to
               maplist(=(0),Us).             % undecided, else 5^k guesses
grounds([U|Us]) :- permute([2,1,0,-1,-2],Ps), member(U,Ps),  % unbiased: nfr2
                   grounds(Us).                              % tried 2 first

% ---- compiler ------------------------------------------------------------
% head(K,H,..): K's compiled predicate gets value + threaded state args.
head(K,H,V,S0,S) :- K =.. Ws, append(Ws,[V,S0,S],Zs), H =.. Zs.

softish(makes(_)). softish(breaks(_)). softish(helps(_)).
softish(hurts(_)). softish(and(_)).   softish(or(_)).

hardlit(no(_)) :- !.
hardlit(L)     :- \+ softish(L).
hard(B)        :- maplist(hardlit,B).

wrap(no(X), no(X))  :- !.                    % bare atom in an edge list
wrap(E,     E)      :- softish(E), !.        % is shorthand for makes
wrap(E,     makes(E)).

ref(no(X),    X) :- !.                       % node atoms a body mentions
ref(makes(X), X) :- !.
ref(breaks(X),X) :- !.
ref(helps(X), X) :- !.
ref(hurts(X), X) :- !.
ref(E,        X) :- (E = and(L) ; E = or(L)), !, member(E1,L), ref(E1,X).
ref(X,        X).

compile1(K) :-
  head(K,H,V,S0,S),                          % memo first: cut kills the body
  assertz((H :- believed(K,W,S0,S), !, V = W)),   % clause, so loops share W

  findall(B, (K <-- B), Bs),
  include(hard, Bs, Hs),
  ( Hs = [_|_] -> compile1or(K,Hs) ; compile1and(K,Bs) ).

compile1or(K,Bs) :-                          % any all-hard body: rules win
  head(K,H,2,S0,S),                          % (only ever prove 2); edges on
  assertz((H :- or(Bs,[K-2|S0],S))).         % a mixed head are dead, as nfr2

compile1and(K,Bs) :-                         % edges: V pending for combine;
  append(Bs,Es0), maplist(wrap,Es0,Es),      % a bound-V demand is assumed,
  head(K,H,V,S0,S),                          % never derived (as nfr2's
  assertz((H :- ( var(V)                     % guess clause)
                -> permute(Es,Rs), calls(Rs,Vs,[K-V|S0],S), combine(Vs,V)
                ;  S = [K-V|S0] ))).

stub(K) :- head(K,H,V,S0,S), assertz((H :- maybe(K,V,S0,S))).

% preprocess: run once, after the model is loaded (so discontiguous
% heads see all their bodies). Compiles every head, stubs every
% body atom that has no clauses. After it, the world is closed:
% calling anything else is an existence error -- typos included.
heads(Ks)  :- findall(K, (K <-- _), Ks0), sort(Ks0,Ks).
leaves(Ls) :- heads(Ks),
              findall(X, ((_ <-- B), member(E,B), ref(E,X)), Xs0),
              sort(Xs0,Xs), subtract(Xs,Ks,Ls).

preprocess :- heads(Ks),  maplist(compile1, Ks),
              leaves(Ls), maplist(stub, Ls).

% ---- report --------------------------------------------------------------
% Types are derived from clause shapes, never declared: rule (any
% all-hard body), edge (other clauses), leaf (referenced, no clauses).
type(K,T) :- ( (K <-- B), hard(B) -> T = rule
             ; (K <-- _)          -> T = edge
             ;                       T = leaf ).

types :- heads(Hs), leaves(Ls), append(Hs,Ls,Ks),
         forall(member(T,[rule,edge,leaf]),
                ( findall(K, (member(K,Ks), type(K,T)), Ts),
                  length(Ts,N), format("~w ~w ~q~n",[T,N,Ts]) )).

list_model :- heads(Ks),                     % clause groups per head,
              forall(member(K,Ks),           % a blank line before each
                     ( nl, forall((K <-- B), portray_clause(K <-- B)) )).

% ---- top ----------------------------------------------------------------
abduce(G,As)   :- lit(G,[],A), picks(A,As).
soften(G,V,As) :- call(G,V,[],A), picks(A,As).

picks(A,As) :- msort(A,Ps), findall(X,(member(K-V,Ps),leaf(K,V,X)),As).
leaf(K,_,_)      :- (K <-- _), !, fail.      % derived: hide, show leaves
leaf(K, 2, K)    :- !.
leaf(K,-2, no K) :- !.
leaf(K, V, lab(K,V)).
